### ISSUE TITLE
Dangerous race conditions and non-atomic procedures in luksroot.nix HMAC-SHA1 procedure leads to data loss

### ISSUE THREAD

**Issue by MatthewCroughan** [NixOS/nixpkgs#210329]
### Describe the bug

https://github.com/NixOS/nixpkgs/blob/e39c39072026604eaaa244fc8f28e44c1051e74c/nixos/modules/system/boot/luksroot.nix#L332-L344

Line 333 applies the `new_key` and permanently mutates the luks header on the disk, changing the luks passphrase **before** it stores the $new_salt to persistent `/crypt-storage`. This means that if there is a power outage on line 334, then the user will lose all access to their data, as the new key passphrase cannot be derived, since the salt was never stored on the persistent `/crypt-storage` in time.

We should store the old salt and the new salt in `/crypt-storage` **first**, and only then apply the `luksChangeKey` (`${cschange}` in the luksroot.nix code). This way **manual recovery** is always possible since the old salt and new salt are available. But no matter what we do, it seems like this code will always be destined to lose user data, I don't see a way of solving this race condition, as you always have to do one procedure before/after, and losing power at that critical moment will result in the inability to boot.

There was an unfinished attempt to make renewing the luks key passphrase with `luksChangeKey` optional, which would have prevented this issue if we were to default it to `false` https://github.com/NixOS/nixpkgs/pull/107038

Perhaps instead of using `luksChangeKey`, we could use `luksAddKey` and keep the old challenge available, only removing it when we're sure we've written the new salt. Though some may argue this can reduce security.

I think this code should be outright removed from Nixpkgs, as it is clearly buggy, unsafe, unatomic and can lead to easy data loss, with no clear safe solution. A friend of mine lost data like this, and it seems by looking at some of the commits to this file, others have as well?

**Comment by roberth** [comment]
> But no matter what we do, it seems like this code will always be destined to lose user data

Only lose the ability to boot, right? You did say manual recovery was possible.
Can't we make the boot process detect this scenario and recover automatically? It could be tested by "configuring" the script to insert an `exit 1` or worse at that critical moment.

**Comment by MatthewCroughan** [comment]
@roberth The manual recovery is still a bad setup and I don't think it's much of an improvement. I should have said that it is destined to fail, one way or another. The user would have to boot into a recovery disk, then mount the drive, then `mv /mnt/crypt-storage/old.seed /mnt/crypt-storage/default` every time this happens.

**Comment by MatthewCroughan** [comment]
The only way we can make the boot process detect this scenario is by using `luksAddKey` to have the old key always be valid until rotation is successful, then we can use `lukeRemoveKey` to remove the old salt and challenge. That is entirely possible, but I wonder if some security expert will tell us this is insecure?

**Comment by roberth** [comment]
A witty security expert might tell you that the current implementation is very secure because it makes attackers completely redundant.
However, this isn't quite true, because an attacker might still try to convince you that it was due to a ransomware and ask for bitcoin ;)

On a more serious note, if add + remove isn't secure enough, LUKS has a bigger problem that isn't up to us to fix.
If I may speculate, a possible reason for `luksChangeKey` to exist is the limited number of slots. It'd be a bit bad if you could only change keys at `n-1` utilization.

**Comment by hugolgst** [comment]
@MatthewCroughan any will to pick up the PR?

**Comment by MatthewCroughan** [comment]
@hugolgst Nope, I think this code should be removed from nixpkgs. systemd-initrd with systemd-cryptsetup is something better to use.

**Comment by MatthewCroughan** [comment]
@hugolgst has the code been removed from nixpkgs? If not the issue should remain open.
